walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,57 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.TreeSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> lst = new ArrayList<>();
+    HashSet<String> mandatoryParticipants = new HashSet<>(request.getAttendees());
+    HashSet<String> optionalParticipants = new HashSet<>(request.getOptionalAttendees());
+    TreeSet<TimeRange> busyTimesMandatory = new TreeSet<>(TimeRange.ORDER_BY_START);
+    TreeSet<TimeRange> busyTimesBoth = new TreeSet<>(TimeRange.ORDER_BY_START);
+
+    for (Event it : events){
+      if (eventHasGivenParticipants(it.getAttendees(), optionalParticipants) || eventHasGivenParticipants(it.getAttendees(), mandatoryParticipants))
+        busyTimesBoth.add(it.getWhen());
+      if (!eventHasGivenParticipants(it.getAttendees(), mandatoryParticipants)) // Do not include event if no mandatory participants attend
+        continue;
+      busyTimesMandatory.add(it.getWhen());
+    }
+
+    Collection<TimeRange> availableTimesForBoth = getAvailableTimes(busyTimesBoth, request.getDuration());
+    if (availableTimesForBoth.isEmpty())
+      return getAvailableTimes(busyTimesMandatory, request.getDuration());
+    else return availableTimesForBoth;
+  }
+
+  private Collection<TimeRange> getAvailableTimes(Collection<TimeRange> busyTimes, long duration){
+    if (busyTimes.isEmpty() && duration <= TimeRange.WHOLE_DAY.duration())
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    int last = TimeRange.START_OF_DAY;
+    
+    for (TimeRange it : busyTimes){
+      if (it.start() >= last && it.start() - last >= duration)
+        availableTimes.add(TimeRange.fromStartEnd(last, it.start(), false));
+      if (last< it.end())
+        last = it.end();
+    }
+    if (TimeRange.END_OF_DAY -  last >= duration)
+      availableTimes.add(TimeRange.fromStartEnd(last, TimeRange.END_OF_DAY, true));
```

Comment:
The definition of END_OF_DAY or of the range is confusing here. For time ranges, `end` is the first minute that is not included in the range -- why is END_OF_DAY included in the last interval then?

Reply:
I just saw in their tests that true should be set for END_OF_DAY

Reply:
Ok.